dali/operators/math/expressions/broadcasting.cc

```diff
@@ -274,33 +274,109 @@ void ExpandToNDims(TensorShape<> &sh, int ndim) {
   sh = shape_cat(TensorShape<>(std::vector<int64_t>(ndim - sh.sample_dim(), 1)), sh);
 }
 
-void SimplifyShapesForBroadcasting(TensorShape<>& lhs, TensorShape<> &rhs) {
-  // First, if needed expand dimensions
-  int full_ndim = std::max(lhs.sample_dim(), rhs.sample_dim());
-  if (lhs.sample_dim() != rhs.sample_dim()) {
-    ExpandToNDims(lhs, full_ndim);
-    ExpandToNDims(rhs, full_ndim);
-  }
+SmallVector<std::pair<int, int>, 5> SimplifyShapesForBroadcasting(span<TensorShape<> *> shapes,
+                                                                  bool modify_shapes) {
+  int n = shapes.size();
+  SmallVector<TensorShape<>, 3> outs;
+  outs.resize(n);
+
+  int ndim = shapes[0]->size();
+  for (int i = 1; i < n; i++)
+    if (static_cast<int>(shapes[i]->size()) > ndim)
+      ndim = shapes[i]->size();
+
+  auto get = [&](int shape, int dim) -> int64_t {
+    auto &s = *shapes[shape];
+    dim -= ndim - s.size();  // add leading unit dims
+    if (dim < 0)
+      return 1;  // implicit unit dim
+    return s[dim];
+  };
+
+  auto should_skip = [&](int d) {
+    for (int i = 0; i < n; i++)
+      if (get(i, d) != 1)
+        return false;
+    return true;
+  };
+
+  SmallVector<int64_t, 3> volumes;
+  volumes.resize(n, 1);
 
-  int i = 0;
-  SmallVector<std::pair<int, int>, 5> group_dims;
-  while (i < full_ndim) {
-    if (lhs[i] != rhs[i]) {
-      i++;
-      continue;
+  SmallVector<std::pair<int, int>, 6> groups;
+  int group_start = 0;
+
+  auto can_collapse = [&](int d) {
+    for (int i = 0; i < n; i++) {
+      bool prev_b = get(i, group_start) == 1;
+      bool curr_b = get(i, d) == 1;
+      if (prev_b != curr_b)
+        return false;
     }
-    int j = i;
-    for (; j < full_ndim; j++) {
-      if (lhs[j] != rhs[j]) break;
+    return true;
+  };
+
+  auto add_group = [&](int d) {
+    if (groups.empty())  // there were leading degenerate dims
+      groups.emplace_back(0, d);
+    else
+      groups.emplace_back(group_start, d);
+    group_start = d;
+    for (int i = 0; i < n; i++) {
+      outs[i].shape.push_back(volumes[i]);
+      volumes[i] = get(i, d);
     }
-    if (i < j) {
-      group_dims.emplace_back(i, j - i);
+  };
+
+  int d = 0;
+  for (; d < ndim; d++) {
+    if (!should_skip(d))
+      break;
+  }
+
+  if (d < ndim) {
+    for (int i = 0; i < n; i++)
+      volumes[i] *= get(i, d);
+
+    for (d++; d < ndim; d++) {
+      if (should_skip(d))
+        continue;
+      if (can_collapse(d)) {
+        for (int i = 0; i < n; i++)
+          volumes[i] *= get(i, d);
+        continue;
+      }
+
+      add_group(d);
+    }
+    add_group(d);
+
+    if (modify_shapes) {
+      for (int i = 0; i < n; i++) {
+        *shapes[i] = outs[i];
+      }
```

Comment:
This should be outside the `if (d < ndim) {` - otherwise the shapes remain unsimplified in a totally degenerate case.